### DISCUSSION TITLE
Water Slides!

### DISCUSSION THREAD

**Discussion by Shima33** [OpenRCT2/OpenRCT2#15746] [Feature requests]
I think having water slides in RCT2 would be really cool. You could re-use the dinghy slide sprites for the track and have it where it is the only coaster that can end without having to connect back to the station if there is enough water in front of the end of the track.

**Comment by ocalhoun6** [comment]
"Having enough water at the end" would be problematic because you'd need new code (and some weird behavior) to get the peeps out of the water and back onto a path somehow.

_But_ I do think there's a good way to overcome this: require the water slide track to end in some special 'splashdown' pieces -- made from splash boat track. A few pieces of flat splash boat track, then a splash boat station as the exit station. You might want to make *one* custom track piece to make a nice transition between dingy slide --> splash boat track while still looking good and realistic.

However, this new ride type would probably require some special code just for it, to handle the ride vehicles disappearing at the exit and magically appearing at the entrance. That's not something any ride in RCT2 does in vanilla, so it would require some custom code made from scratch to govern that behavior. It would also need custom code and custom error messages to make sure the first station is entry-only and the second station is exit-only.

It could also be interesting to make this the only vertically moving ride that doesn't have access to a chain lift or self-powered vehicles. So you could still have uphill sections, but only if the ride has enough momentum to get over those hills naturally.

All that said, I doubt anybody is going to make this, since it's already possible to approximate it with cheats. (All you really need is to make the return track invisible or hidden underground/behind scenery.) Still, it would be cool to have this as a ride type.

**Comment by Ryder17z** [comment]
> _But_ I do think there's a good way to overcome this: require the water slide track to end in some special 'splashdown' pieces -- made from splash boat track.

Logflume could alternatively be used for this.

> However, this new ride type would probably require some special code just for it, to handle the ride vehicles disappearing at the exit and magically appearing at the entrance. That's not something any ride in RCT2 does in vanilla, so it would require some custom code made from scratch to govern that behavior. It would also need custom code and custom error messages to make sure the first station is entry-only and the second station is exit-only.

[Marcel Vos did make teleporting gokarts](https://www.youtube.com/watch?v=nx6Q-bioFW4), so it is possible to make a vehicle skip track pieces, but it's not exactly a good way to do it.


But there is a way to currently do this;
![image](https://user-images.githubusercontent.com/2000703/122115979-516ad600-ce25-11eb-92b0-5aeaed7d2657.png)
I don't know who made the SLDBODY.DAT file (left ride), the right ride file is called AE-WSLID.DAT

The biggest issue getting the speed to look right with either of these. For the return track;  You could hide it using tunnels. (i have hidden the return track using the tile inspector here)

**Comment by spacek531** [comment]
Teleporting a car when it reaches the end of a track is relatively simple, but teleporting a car when it reaches the end of the track while checking for collision on the track element it will teleport to is more challenging. Here's some ideas I had on how it could work.

In UpdateTrackMotionForwardsGetNewTrack and UpdateTrackMotionBackwardsGetNewTrack,  if the vehicle in question is the front vehicle relative to the direction of motion and is allowed to teleport and the new track is a terminus:

Seek backwards relative to its motion until it finds the other terminus. Create an imaginary track piece based on the properties of the other terminus (the default track piece when you build a ride). Create an imaginary car and set it as the head of the train relative to its motion, make it the same size as the car it is standing in for, and make it not affect the acceleration of the train.

Place the imaginary car on the track so that it is in the same distance from the end of the track as the first car is from the end of its track when going forward or place it the same distance from the start of the track as the last car is from the start of its track if going backwards. Adding something to the TED might help with this.

Finally, check for collision with the head or tail end of the next train, or move to to the next car if it determines that it encountered an imaginary car.

When the imaginary car calls UpdateTrackMotionForwardsGetNewTrack or UpdateMotionBackwardsGetNewTrack, the head or tail car is teleported to it and the imaginary car gets removed. Every subsequent car that calls UpdateTrackMotionForwardsGetNewTrack or UpdateMotionBackwardsGetNewTrack simply teleports to the desired track since the imaginary car has already cleared the way.

**Comment by duncanspumpkin** [comment]
I feel a better way would be to create a new track piece that is a vehicle spawner station. Make it so that it can spawn a vehicle after station is clear. Would need to modify the ride structure a bit to handle it.

**Comment by spacek531** [comment]
It would definitely be simpler to produce and consume trains. I think a new piece may be unnecessary since a new ride mode is necessary anyway. The ride can consume vehicles when they unload and produce new ones at the original station.

Your proposal would solve the water slide in a much simpler way that what I've proposed, but I think general teleportation would be pretty cool.

**Comment by duncanspumpkin** [comment]
A teleporter I think most would expect momentum to be conserved so it feels like a completely different feature imho

**Comment by spacek531** [comment]
I also expect momentum to be conserved. I think my proposal will conserve it. As far as if momentum should be conserved for a water slide, I could see it being either way.

**Comment by duncanspumpkin** [comment]
My point is that a water slide you would definitely not want conserved momentum as your dingy might fly off the station platform.

**Comment by spacek531** [comment]
If there is a station at the bottom, it will move at station speeds when it reaches the end of that station to teleport back. So I don't see momentum conservation as a bad thing.

**Comment by spacek531** [comment]
There's some quirks with how RCT2 handles track piece length: https://github.com/OpenRCT2/OpenRCT2/wiki/Track-Element-Lengths

It's not actually easy to know when the end of the track is because tracks actually change length depending on which way they are rotated.

This will make it more difficult to make a conservative teleporting feature.

**Comment by J-PIE-314** [comment]
PR #19681 is meant to implement this feature.
